### PR TITLE
Eliminates roundtrips between the browser and the renderer.

### PR DESCRIPTION
**Pull request by szilardszaloki**
Resolves https://github.com/brave/brave-browser/issues/36168.

Component PR: https://github.com/brave/playlist-component/pull/40

- moves hardcoded scripts from the browser to component `==>` every Playlist-related script is now in the component (see component PR above)
- eliminates roundtrips between the browser and the renderer (details [here](https://bravesoftware.slack.com/archives/CKJCXTP9B/p1706651504711089)) `==>` browser passes scripts to `PlaylistRenderFrameObserver` as part of `RenderFrame` startup
- alleviates the need for using Blink preferences for browser-renderer IPC (`force_cosmetic_filtering` will also be removed when we add `MediaSource` API support)
- breaking out a `PlaylistBackgroundWebContentsHelper` from `PlaylistTabHelper` for background `WebContents`
- simplifies architecture around script injection

TODO:
- [x] update `DEPS` once component PR is merged

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/playlist/browser/playlist_background_webcontents_helper.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      content::NavigationHandle* navigation_handle) override;
+
+ private:
+  friend class content::WebContentsUserData<
+      PlaylistBackgroundWebContentsHelper>;
+
+  PlaylistBackgroundWebContentsHelper(
+      content::WebContents* web_contents,
+      const std::string& media_source_api_suppressor,
+      const std::string& media_detector);
```
Comment (sangwoo108) [inline components/playlist/browser/playlist_background_webcontents_helper.h:42]: Nice refactoring. How about moving `fake UA` stuff here together?

File: components/playlist/browser/playlist_background_webcontents_helper.cc [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+void PlaylistBackgroundWebContentsHelper::ReadyToCommitNavigation(
+    content::NavigationHandle* navigation_handle) {
+  DVLOG(2) << __FUNCTION__;
+
+  DCHECK(navigation_handle);
+  if (!navigation_handle->IsInPrimaryMainFrame()) {
+    return;
+  }
+
+  const GURL url = navigation_handle->GetWebContents()->GetVisibleURL();
```
Comment (sangwoo108) [inline components/playlist/browser/playlist_background_webcontents_helper.cc:29]: The latest version is using `GetLastCommittedURL()` instead of `GetVisibleURL()` following https://github.com/brave/brave-core/pull/21332#discussion_r1487668333

File: components/playlist/browser/playlist_tab_helper.cc [hunk 3]
```diff
@@ -181,14 +183,26 @@ void PlaylistTabHelper::ExtractMediaFromBackgroundWebContents(
...
+void PlaylistTabHelper::ReadyToCommitNavigation(
+    content::NavigationHandle* navigation_handle) {
+  DVLOG(2) << __FUNCTION__;
+
+  if (!navigation_handle->IsInPrimaryMainFrame()) {
+    return;
+  }
+
+  const GURL url = navigation_handle->GetWebContents()->GetVisibleURL();
+  if (url.SchemeIs(url::kAboutScheme)) {
```
Comment (sangwoo108) [inline components/playlist/browser/playlist_tab_helper.cc:195]: ditto

File: components/playlist/browser/playlist_background_webcontents_helper.cc [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  if (url.SchemeIs(url::kAboutScheme)) {
```
[the remark this answers is not in the document]
Comment (szilardszaloki) [reply components/playlist/browser/playlist_background_webcontents_helper.cc:30]: Yup, let's go with `SchemeIsHTTPOrHTTPS()`! 👍🏻

File: components/playlist/renderer/playlist_render_frame_observer.h [hunk 5]
```diff
@@ -6,41 +6,65 @@
...
 
-  void HideMediaSourceAPI() const;
-  void InstallMediaDetector();
+  const mojo::Remote<playlist::mojom::PlaylistMediaHandler>& GetMediaHandler();
 
-  void OnMediaUpdated(const std::string& page_url);
+  void Inject(const std::string& script_text,
+              v8::Local<v8::Context> context,
+              std::vector<v8::Local<v8::Value>> args = {}) const;
+  void OnMediaDetected(base::Value media);
```
[the remark this answers is not in the document]
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.h:59]: Hmm, is this a Chromium C++ guideline?

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 6]
```diff
@@ -35,137 +66,108 @@ void PlaylistRenderFrameObserver::OnDestruct() {
...
-          context->GetIsolate(),
-          base::BindRepeating(&PlaylistRenderFrameObserver::OnMediaUpdated,
-                              weak_ptr_factory_.GetWeakPtr()))
-          ->GetFunction(context)
-          .ToLocalChecked();
-  v8::Local<v8::Value> arg = on_media_updated.As<v8::Value>();
-
-  std::ignore = function->Call(context, context->Global(), 1, &arg);
+  std::ignore = function->Call(context, context->Global(), args.size(),
+                               args.empty() ? nullptr : args.data());
```
[the remark this answers is not in the document]
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:164]: This ternary is necessary, as the C++ standard doesn't explicitly say that `data()` is `nullptr` on an empty `std::vector<>`.

File: components/playlist/browser/playlist_background_webcontents_helper.h (hunk 1 above)
Comment (szilardszaloki) [reply components/playlist/browser/playlist_background_webcontents_helper.h:42]: From what I can see, we have to configure user agent-related settings on the `WebContents`, the `NavigationController`, and on the `LoadURLParams`. The latter wouldn't be available in `PlaylistBackgroundWebContentsHelper`. It would be nice to have all three of them occurring in the same function, and actually, I have a refactor in the works, that already has that.

File: components/playlist/renderer/playlist_render_frame_observer.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.h:59]: I can't find any written docs for it, so it's just from my experience. You can leave it as is if you want.

File: components/playlist/renderer/playlist_render_frame_observer.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.cc:164]: Yeah, in addition to that, I was wondering passing valid pointer and 0 size could cause crash or something. (But I do like the code as is 😄)

File: components/playlist/browser/playlist_background_webcontents_helper.cc (hunk 2 above)
Comment (szilardszaloki) [reply components/playlist/browser/playlist_background_webcontents_helper.cc:29]: `GetLastCommittedURL()` is not the URL we want  in `ReadyToCommitNavigation()` — this navigation hasn't committed yet.

File: components/playlist/browser/playlist_tab_helper.cc (hunk 3 above)
Comment (szilardszaloki) [reply components/playlist/browser/playlist_tab_helper.cc:195]: Same here.

File: components/playlist/renderer/playlist_render_frame_observer.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:164]: in practice, I'd be surprised if it mattered, but the correct usage requires `nullptr` when `argc == 0`, and unfortunately, we don't get it from the standard (https://eel.is/c++draft/vector.data). [CppReference](https://en.cppreference.com/w/cpp/container/vector/data) even calls it out in `Notes`.

File: components/playlist/browser/playlist_background_webcontents_helper.cc (hunk 2 above)
Comment (sangwoo108) [reply components/playlist/browser/playlist_background_webcontents_helper.cc:29]: Ah ha, I see, it's still a pending entry.

File: components/playlist/renderer/playlist_render_frame_observer.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.h]: Yeah, unless there's a Chromium guideline for that we need to follow, I'd prefer it that way. Having the overridden functions listed first makes more sense, as it's analogous to how object layout works (inherited members come first).

[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.h]: I see where you're coming from. Good to know 👍

**Review by mkarolin — APPROVED** [review]
`DEPS`, `chromium_src`, `strings` LGTM

File: components/playlist/browser/playlist_service.cc [hunk 7]
```diff
@@ -364,11 +366,7 @@ void PlaylistService::NotifyPlaylistChanged(mojom::PlaylistEvent playlist_event,
 
 void PlaylistService::NotifyMediaFilesUpdated(
     const GURL& url,
-    std::vector<mojom::PlaylistItemPtr> items) {
-  if (items.empty()) {
```
Comment (bridiver) [inline components/playlist/browser/playlist_service.cc:368]: why was this moved to the observers? That seems to make it more fragile, no?

Comment (szilardszaloki) [reply components/playlist/browser/playlist_service.cc:368]: This check was moved here as an intermediary step (was easier to adjust tests). It's being removed everywhere in my current refactoring work, as the renderer shouldn't even send empty `base::Value::List`s.

File: components/playlist/browser/media_detector_component_manager.cc [hunk 8]
```diff
@@ -138,7 +148,11 @@ void MediaDetectorComponentManager::OnGetScripts(
     return;
   }
 
-  DCHECK(script_map.count(GetBaseScriptName()));
+  CHECK(script_map.contains(GetMediaSourceAPISuppressorScriptName()));
```
Comment (bridiver) [inline components/playlist/browser/media_detector_component_manager.cc:151]: I'm not sure if this should be a `CHECK`? Do we want to crash if there's some mistake in the crx?

File: components/playlist/browser/playlist_background_webcontents_helper.cc (hunk 2 above)
Comment (bridiver) [reply components/playlist/browser/playlist_background_webcontents_helper.cc:28]: Those are both incorrect. You should be using `navigation_handle->GetURL()` to get the url that is being navigated to. If there is any other code outside of this using `GetLastCommittedURL` or `GetVisibleURL` instead of the navigation handle it should also be fixed in a separate PR. 
https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/navigation_handle.h;l=100?q=navigation_handle.h&ss=chromium

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 9]
```diff
@@ -35,137 +66,108 @@ void PlaylistRenderFrameObserver::OnDestruct() {
   delete this;
 }
 
-bool PlaylistRenderFrameObserver::EnsureConnectedToMediaHandler() {
-  if (!media_handler_.is_bound()) {
+void PlaylistRenderFrameObserver::AddMediaSourceAPISuppressor(
+    const std::string& media_source_api_suppressor) {
+  DVLOG(2) << __FUNCTION__;
```
Comment (bridiver) [inline components/playlist/renderer/playlist_render_frame_observer.cc:71]: are all these DVLOGs really necessary? Chromium docs state that most logging should be removed prior to merge

File: components/playlist/browser/media_detector_component_manager.cc (hunk 8 above)
Comment (szilardszaloki) [reply components/playlist/browser/media_detector_component_manager.cc:151]: `script_map` is non-empty at this point, so I think so.

File: components/playlist/renderer/playlist_render_frame_observer.cc (hunk 9 above)
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:71]: They do come in handy sometimes, and are just `DVLOG`s. I also see Chromium doing the same (~10k instances).

File: components/playlist/browser/playlist_background_webcontents_helper.cc (hunk 2 above)
Comment (szilardszaloki) [reply components/playlist/browser/playlist_background_webcontents_helper.cc]: Fair enough — fixed.

File: components/playlist/browser/media_detector_component_manager.cc (hunk 8 above)
Comment (bridiver) [reply components/playlist/browser/media_detector_component_manager.cc:151]: but what I'm saying is if there's some screwup with the crx and it doesn't contain the file we expect, crashing doesn't seem ideal. I think we would just bail out of the update instead?

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 10]
```diff
@@ -35,137 +66,108 @@ void PlaylistRenderFrameObserver::OnDestruct() {
...
+    v8::Isolate* isolate = blink::MainThreadIsolate();
+    v8::Isolate::Scope isolate_scope(isolate);
+    v8::HandleScope handle_scope(isolate);
+
+    v8::Local<v8::Context> context =
+#if !BUILDFLAG(IS_ANDROID)
+        render_frame()->GetWebFrame()->GetScriptContextFromWorldId(
+            isolate, isolated_world_id_);
+#else
+        render_frame()->GetWebFrame()->MainWorldScriptContext();
```
Comment (bridiver) [inline components/playlist/renderer/playlist_render_frame_observer.cc:129]: Why are we running this in the main world? Didn't it previously execute in an isolated world? Running it in the main world is potentially dangerous because of prototype pollution

Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:129]: There's no change in behavior (yet): Android injects into main, desktop injects into `ISOLATED_WORLD_ID_BRAVE_INTERNAL`. IIRC, that's because Android needs https://github.com/brave/playlist-component/blob/main/youtube.com.js to extract thumbnails (which uses `window.ytplayer...`, hence needs main world access), as the necessary OG `<meta>` tags (which is how desktop does it) are not present on Android.

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 11]
```diff
@@ -5,28 +5,59 @@
...
 #include "third_party/blink/public/web/blink.h"
 #include "third_party/blink/public/web/web_local_frame.h"
-#include "third_party/blink/public/web/web_script_source.h"
-#include "url/gurl.h"
 #include "v8/include/v8.h"
 
+namespace gin {
+
+template <>
+struct Converter<base::Value> {
```
Comment (bridiver) [inline components/playlist/renderer/playlist_render_frame_observer.cc:26]: there is already a converter for `base::Value`

Comment (bridiver) [reply components/playlist/renderer/playlist_render_frame_observer.cc:26]: nvm, I guess electron added this and I was thinking it was a chromium default converter

File: components/playlist/renderer/playlist_render_frame_observer.cc (hunk 10 above)
Comment (bridiver) [reply components/playlist/renderer/playlist_render_frame_observer.cc:129]: I believe you have already filed an issue for this?

File: components/playlist/browser/media_detector_component_manager.cc (hunk 8 above)
Comment (szilardszaloki) [reply components/playlist/browser/media_detector_component_manager.cc:151]: Creating a follow-up issue for that. Also, these changes don't change the current behavior, as `std::map<>::at()` throws `std::out_of_range` if the specific key is not in the map.

File: components/playlist/renderer/playlist_render_frame_observer.cc (hunk 10 above)
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:129]: Not yet — raising now.

Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:129]: https://github.com/brave/brave-browser/issues/36443

File: components/playlist/browser/media_detector_component_manager.cc (hunk 8 above)
Comment (szilardszaloki) [reply components/playlist/browser/media_detector_component_manager.cc:151]: https://github.com/brave/brave-browser/issues/36445

**Review by bridiver — APPROVED** [review]
